File: src/main_window.cpp
#include "main_window.hpp"
#include <algorithm>
#include <exception>
#include <iterator>
#include "core/provider_registry.hpp"

namespace {
constexpr int categories_label_id = 101;
constexpr int categories_list_id = 102;
constexpr int properties_label_id = 103;
constexpr int properties_list_id = 104;
constexpr int menu_refresh_id = 201;
constexpr int menu_exit_id = 202;
constexpr int menu_copy_id = 203;

// Builds a minimal in-memory DLGTEMPLATE with zero template-defined items; every control is created by hand in on_init_dialog. Using the predefined dialog class (windowClass = 0) is what gets us free accessible-name association between each STATIC label and the control after it, plus Tab-key navigation via IsDialogMessage - neither works for a plain CreateWindowEx window. Only the outer size is approximate here; create() overrides it in real pixels via SetWindowPos right after creation.
std::vector<BYTE> build_dialog_template(const std::wstring& title) {
	std::vector<BYTE> buffer;
	auto append_word = [&](WORD w) { auto* p = reinterpret_cast<BYTE*>(&w); buffer.insert(buffer.end(), p, p + sizeof(WORD)); };
	auto append_dword = [&](DWORD d) { auto* p = reinterpret_cast<BYTE*>(&d); buffer.insert(buffer.end(), p, p + sizeof(DWORD)); };
	auto append_wstring = [&](const std::wstring& s) { auto* p = reinterpret_cast<const BYTE*>(s.c_str()); buffer.insert(buffer.end(), p, p + (s.size() + 1) * sizeof(wchar_t)); };
	const DWORD style = DS_SETFONT | WS_POPUP | WS_CAPTION | WS_SYSMENU | WS_THICKFRAME | WS_MINIMIZEBOX | WS_MAXIMIZEBOX | WS_CLIPCHILDREN;
	append_dword(style);
	append_dword(0);
	append_word(0);
	append_word(0);
	append_word(0);
	append_word(300);
	append_word(200);
	append_word(0);
	append_word(0);
	append_wstring(title);
	append_word(8);
	append_wstring(L"MS Shell Dlg");
	return buffer;
}
}  // namespace

main_window::main_window(HINSTANCE instance) : instance_(instance) {}

HWND main_window::create() {
	std::vector<BYTE> tmpl = build_dialog_template(L"System Information");
	hwnd_ = CreateDialogIndirectParamW(instance_, reinterpret_cast<LPCDLGTEMPLATE>(tmpl.data()), nullptr, dialog_proc, reinterpret_cast<LPARAM>(this));
	if (hwnd_) {
		const int width = 900;
		const int height = 600;
		const int screen_width = GetSystemMetrics(SM_CXSCREEN);
		const int screen_height = GetSystemMetrics(SM_CYSCREEN);
		SetWindowPos(hwnd_, nullptr, (screen_width - width) / 2, (screen_height - height) / 2, width, height, SWP_NOZORDER);
	}
	return hwnd_;
}

INT_PTR CALLBACK main_window::dialog_proc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam) {
	main_window* self;
	if (msg == WM_INITDIALOG) {
		self = reinterpret_cast<main_window*>(lparam);
		SetWindowLongPtrW(hwnd, GWLP_USERDATA, reinterpret_cast<LONG_PTR>(self));
		self->hwnd_ = hwnd;
	} else {
		self = reinterpret_cast<main_window*>(GetWindowLongPtrW(hwnd, GWLP_USERDATA));
	}
	if (self) return self->handle_message(hwnd, msg, wparam, lparam);
	return FALSE;
}

INT_PTR main_window::handle_message(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam) {
	switch (msg) {
		case WM_INITDIALOG: on_init_dialog(hwnd); return TRUE;
		case WM_SIZE: on_size(LOWORD(lparam), HIWORD(lparam)); return TRUE;
		case WM_COMMAND: on_command(wparam, lparam); return TRUE;
		case WM_CLOSE: DestroyWindow(hwnd); return TRUE;
		case WM_DESTROY: PostQuitMessage(0); return TRUE;
		default: return FALSE;
	}
}

void main_window::on_init_dialog(HWND hwnd) {
	HFONT font = static_cast<HFONT>(GetStockObject(DEFAULT_GUI_FONT));
	categories_label_ = CreateWindowExW(0, L"STATIC", L"Categories", WS_CHILD | WS_VISIBLE | WS_GROUP, 0, 0, 0, 0, hwnd, reinterpret_cast<HMENU>(static_cast<INT_PTR>(categories_label_id)), instance_, nullptr);
	categories_list_ = CreateWindowExW(WS_EX_CLIENTEDGE, L"LISTBOX", nullptr, WS_CHILD | WS_VISIBLE | WS_VSCROLL | WS_TABSTOP | LBS_NOTIFY | LBS_HASSTRINGS, 0, 0, 0, 0, hwnd, reinterpret_cast<HMENU>(static_cast<INT_PTR>(categories_list_id)), instance_, nullptr);
	properties_label_ = CreateWindowExW(0, L"STATIC", L"Properties", WS_CHILD | WS_VISIBLE | WS_GROUP, 0, 0, 0, 0, hwnd, reinterpret_cast<HMENU>(static_cast<INT_PTR>(properties_label_id)), instance_, nullptr);
	properties_list_ = CreateWindowExW(WS_EX_CLIENTEDGE, L"LISTBOX", nullptr, WS_CHILD | WS_VISIBLE | WS_VSCROLL | WS_TABSTOP | LBS_NOTIFY | LBS_HASSTRINGS, 0, 0, 0, 0, hwnd, reinterpret_cast<HMENU>(static_cast<INT_PTR>(properties_list_id)), instance_, nullptr);
	for (HWND control : {categories_label_, categories_list_, properties_label_, properties_list_}) SendMessageW(control, WM_SETFONT, reinterpret_cast<WPARAM>(font), TRUE);
	HMENU file_menu = CreatePopupMenu();
	AppendMenuW(file_menu, MF_STRING, menu_refresh_id, L"&Refresh\tF5");
	AppendMenuW(file_menu, MF_SEPARATOR, 0, nullptr);
	AppendMenuW(file_menu, MF_STRING, menu_exit_id, L"E&xit");
	HMENU edit_menu = CreatePopupMenu();
	AppendMenuW(edit_menu, MF_STRING, menu_copy_id, L"&Copy\tCtrl+C");
	HMENU menu_bar = CreateMenu();
	AppendMenuW(menu_bar, MF_POPUP, reinterpret_cast<UINT_PTR>(file_menu), L"&File");
	AppendMenuW(menu_bar, MF_POPUP, reinterpret_cast<UINT_PTR>(edit_menu), L"&Edit");
	SetMenu(hwnd, menu_bar);
	ACCEL accels[] = {{FVIRTKEY, VK_F5, menu_refresh_id}, {FVIRTKEY | FCONTROL, L'C', menu_copy_id}};
	accelerators_ = CreateAcceleratorTable(accels, static_cast<int>(std::size(accels)));
	refresh_categories();
	SetFocus(categories_list_);
}

void main_window::on_size(int width, int height) {
	if (!categories_list_) return;
	const int margin = 10;
	const int label_height = 18;
	const int half_width = (width - margin * 3) / 2;
	const int list_top = margin + label_height;
	const int list_height = height - list_top - margin;
	const int right_x = margin * 2 + half_width;
	MoveWindow(categories_label_, margin, margin, half_width, label_height, TRUE);
	MoveWindow(categories_list_, margin, list_top, half_width, list_height, TRUE);
	MoveWindow(properties_label_, right_x, margin, half_width, label_height, TRUE);
	MoveWindow(properties_list_, right_x, list_top, half_width, list_height, TRUE);
}

void main_window::on_command(WPARAM wparam, LPARAM lparam) {
	int id = LOWORD(wparam);
	if (lparam != 0) {
		if (id == categories_list_id && HIWORD(wparam) == LBN_SELCHANGE) {
			int index = static_cast<int>(SendMessageW(categories_list_, LB_GETCURSEL, 0, 0));
			populate_properties(index);
		}
		return;
	}
	switch (id) {
		case menu_refresh_id: refresh_categories(); break;
		case menu_exit_id: DestroyWindow(hwnd_); break;
		case menu_copy_id: copy_selected_property(); break;
		// IsDialogMessage turns an unhandled Escape keypress into WM_COMMAND(IDCANCEL); treat it as "close the window".
		case IDCANCEL: DestroyWindow(hwnd_); break;
	}
}

void main_window::refresh_categories() {
	SendMessageW(categories_list_, LB_RESETCONTENT, 0, 0);
	SendMessageW(properties_list_, LB_RESETCONTENT, 0, 0);
	std::vector<category_item> collected;
	for (auto& provider : provider_registry::instance().create_all()) {
		try {
			for (auto& item : provider->get_items()) collected.push_back(std::move(item));
		} catch (const std::exception&) {
			// One provider failing (e.g. a WMI class unavailable on this machine) must not take down the rest of the categories.
		}
	}
	std::sort(collected.begin(), collected.end(), [](const category_item& a, const category_item& b) { return a.label < b.label; });
	items_ = std::move(collected);
	for (const auto& item : items_) SendMessageW(categories_list_, LB_ADDSTRING, 0, reinterpret_cast<LPARAM>(item.label.c_str()));
	// Pre-select the first category (and its first property below) so a screen reader user tabbing in immediately hears something useful, instead of having to press Down arrow once just to get a selection.
	if (!items_.empty()) {
		SendMessageW(categories_list_, LB_SETCURSEL, 0, 0);
		populate_properties(0);
	}
}

void main_window::populate_properties(int category_index) {
	SendMessageW(properties_list_, LB_RESETCONTENT, 0, 0);
	if (category_index < 0 || static_cast<size_t>(category_index) >= items_.size()) return;
	for (const auto& prop : items_[category_index].properties) {
		if (prop.value.empty()) continue;
		std::wstring line = prop.name + L": " + prop.value;
		SendMessageW(properties_list_, LB_ADDSTRING, 0, reinterpret_cast<LPARAM>(line.c_str()));
	}
	if (SendMessageW(properties_list_, LB_GETCOUNT, 0, 0) > 0) SendMessageW(properties_list_, LB_SETCURSEL, 0, 0);
}

void main_window::copy_selected_property() {
	if (GetFocus() != properties_list_) return;
	int index = static_cast<int>(SendMessageW(properties_list_, LB_GETCURSEL, 0, 0));
	if (index == LB_ERR) return;
	int len = static_cast<int>(SendMessageW(properties_list_, LB_GETTEXTLEN, index, 0));
	if (len == LB_ERR) return;
	std::wstring text(static_cast<size_t>(len) + 1, L'\0');
	SendMessageW(properties_list_, LB_GETTEXT, index, reinterpret_cast<LPARAM>(text.data()));
	text.resize(static_cast<size_t>(len));
	if (!OpenClipboard(hwnd_)) return;
	EmptyClipboard();
	HGLOBAL mem = GlobalAlloc(GMEM_MOVEABLE, (text.size() + 1) * sizeof(wchar_t));
	if (mem) {
		wchar_t* dest = static_cast<wchar_t*>(GlobalLock(mem));
		if (dest) {
			wcscpy_s(dest, text.size() + 1, text.c_str());
			GlobalUnlock(mem);
			if (!SetClipboardData(CF_UNICODETEXT, mem)) GlobalFree(mem);
		} else {
			GlobalFree(mem);
		}
	}
	CloseClipboard();
}

File: src/main_window.zig
const std = @import("std");
const win32 = @import("win32.zig");
const provider_registry = @import("core/provider_registry.zig");
const category_item = @import("core/category_item.zig");
const CategoryItem = category_item.CategoryItem;

const categories_label_id: usize = 101;
const categories_list_id: usize = 102;
const properties_label_id: usize = 103;
const properties_list_id: usize = 104;
const menu_refresh_id: usize = 201;
const menu_exit_id: usize = 202;
const menu_copy_id: usize = 203;

fn idMenu(id: usize) win32.HMENU {
	return @ptrFromInt(id);
}

fn ptrToLparam(ptr: anytype) win32.LPARAM {
	return @bitCast(@intFromPtr(ptr));
}

fn lparamToPtr(comptime T: type, lparam: win32.LPARAM) T {
	return @ptrFromInt(@as(usize, @bitCast(lparam)));
}

fn toWide(allocator: std.mem.Allocator, s: []const u8) ![:0]u16 {
	return std.unicode.utf8ToUtf16LeAllocZ(allocator, s);
}

/// Builds a minimal in memory DLGTEMPLATE with zero template defined items, every control is created by hand in onInitDialog. Using the predefined dialog class (windowClass = 0) is what gets us free accessible name association between each STATIC label and the control after it, plus Tab key navigation via IsDialogMessage, neither works for a plain CreateWindowEx window. Only the outer size is approximate here, create() overrides it in real pixels via SetWindowPos right after creation.
fn buildDialogTemplate(allocator: std.mem.Allocator, title: [:0]const u16) ![]u8 {
	var buffer = std.ArrayList(u8).empty;
	const style: u32 = win32.WS_POPUP | win32.WS_CAPTION | win32.WS_SYSMENU | win32.WS_THICKFRAME | win32.WS_MINIMIZEBOX | win32.WS_MAXIMIZEBOX | win32.WS_CLIPCHILDREN | win32.DS_SETFONT;
	try buffer.appendSlice(allocator, std.mem.asBytes(&style));
	try buffer.appendSlice(allocator, std.mem.asBytes(&@as(u32, 0)));
	try buffer.appendSlice(allocator, std.mem.asBytes(&@as(u16, 0)));
	try buffer.appendSlice(allocator, std.mem.asBytes(&@as(u16, 0)));
	try buffer.appendSlice(allocator, std.mem.asBytes(&@as(u16, 0)));
	try buffer.appendSlice(allocator, std.mem.asBytes(&@as(u16, 300)));
	try buffer.appendSlice(allocator, std.mem.asBytes(&@as(u16, 200)));
	try buffer.appendSlice(allocator, std.mem.asBytes(&@as(u16, 0)));
	try buffer.appendSlice(allocator, std.mem.asBytes(&@as(u16, 0)));
	try buffer.appendSlice(allocator, std.mem.sliceAsBytes(title[0 .. title.len + 1]));
	try buffer.appendSlice(allocator, std.mem.asBytes(&@as(u16, 8)));
	const font_name = std.unicode.utf8ToUtf16LeStringLiteral("MS Shell Dlg");
	try buffer.appendSlice(allocator, std.mem.sliceAsBytes(font_name[0 .. font_name.len + 1]));
	return buffer.toOwnedSlice(allocator);
}

pub const MainWindow = struct {
	instance: win32.HINSTANCE,
	hwnd: win32.HWND = null,
	categories_label: win32.HWND = null,
	categories_list: win32.HWND = null,
	properties_label: win32.HWND = null,
	properties_list: win32.HWND = null,
	accelerators: win32.HACCEL = null,
	items: []const CategoryItem = &.{},
	arena: std.heap.ArenaAllocator,

	pub fn init(instance: win32.HINSTANCE, backing_allocator: std.mem.Allocator) MainWindow {
		return .{ .instance = instance, .arena = std.heap.ArenaAllocator.init(backing_allocator) };
	}

	pub fn create(self: *MainWindow) !win32.HWND {
		const template = try buildDialogTemplate(std.heap.page_allocator, std.unicode.utf8ToUtf16LeStringLiteral("System Information"));
		defer std.heap.page_allocator.free(template);
		self.hwnd = win32.CreateDialogIndirectParamW(self.instance, template.ptr, null, dialogProc, ptrToLparam(self));
		if (self.hwnd != null) {
			const width: i32 = 900;
			const height: i32 = 600;
			const screen_width = win32.GetSystemMetrics(win32.SM_CXSCREEN);
			const screen_height = win32.GetSystemMetrics(win32.SM_CYSCREEN);
			_ = win32.SetWindowPos(self.hwnd, null, @divTrunc(screen_width - width, 2), @divTrunc(screen_height - height, 2), width, height, win32.SWP_NOZORDER);
		}
		return self.hwnd;
	}

	fn dialogProc(hwnd: win32.HWND, msg: win32.UINT, wparam: win32.WPARAM, lparam: win32.LPARAM) callconv(.c) win32.INT_PTR {
		var self: ?*MainWindow = null;
		if (msg == win32.WM_INITDIALOG) {
			self = lparamToPtr(*MainWindow, lparam);
			_ = win32.SetWindowLongPtrW(hwnd, win32.GWLP_USERDATA, @bitCast(@intFromPtr(self.?)));
			self.?.hwnd = hwnd;
		} else {
			const raw = win32.GetWindowLongPtrW(hwnd, win32.GWLP_USERDATA);
			if (raw != 0) self = @ptrFromInt(@as(usize, @bitCast(raw)));
		}
		if (self) |s| return s.handleMessage(hwnd, msg, wparam, lparam);
		return 0;
	}

	fn handleMessage(self: *MainWindow, hwnd: win32.HWND, msg: win32.UINT, wparam: win32.WPARAM, lparam: win32.LPARAM) win32.INT_PTR {
		switch (msg) {
			win32.WM_INITDIALOG => {
				self.onInitDialog(hwnd);
				return 1;
			},
			win32.WM_SIZE => {
				self.onSize(win32.loword(lparam), win32.hiword(lparam));
				return 1;
			},
			win32.WM_COMMAND => {
				self.onCommand(wparam, lparam);
				return 1;
			},
			win32.WM_CLOSE => {
				_ = win32.DestroyWindow(hwnd);
				return 1;
			},
			win32.WM_DESTROY => {
				win32.PostQuitMessage(0);
				return 1;
			},
			else => return 0,
		}
	}

	fn onInitDialog(self: *MainWindow, hwnd: win32.HWND) void {
		const font = win32.GetStockObject(win32.DEFAULT_GUI_FONT);
		self.categories_label = win32.CreateWindowExW(0, std.unicode.utf8ToUtf16LeStringLiteral("STATIC"), std.unicode.utf8ToUtf16LeStringLiteral("Categories"), win32.WS_CHILD | win32.WS_VISIBLE | win32.WS_GROUP, 0, 0, 0, 0, hwnd, idMenu(categories_label_id), self.instance, null);
		self.categories_list = win32.CreateWindowExW(win32.WS_EX_CLIENTEDGE, std.unicode.utf8ToUtf16LeStringLiteral("LISTBOX"), null, win32.WS_CHILD | win32.WS_VISIBLE | win32.WS_VSCROLL | win32.WS_TABSTOP | win32.LBS_NOTIFY | win32.LBS_HASSTRINGS, 0, 0, 0, 0, hwnd, idMenu(categories_list_id), self.instance, null);
		self.properties_label = win32.CreateWindowExW(0, std.unicode.utf8ToUtf16LeStringLiteral("STATIC"), std.unicode.utf8ToUtf16LeStringLiteral("Properties"), win32.WS_CHILD | win32.WS_VISIBLE | win32.WS_GROUP, 0, 0, 0, 0, hwnd, idMenu(properties_label_id), self.instance, null);
		self.properties_list = win32.CreateWindowExW(win32.WS_EX_CLIENTEDGE, std.unicode.utf8ToUtf16LeStringLiteral("LISTBOX"), null, win32.WS_CHILD | win32.WS_VISIBLE | win32.WS_VSCROLL | win32.WS_TABSTOP | win32.LBS_NOTIFY | win32.LBS_HASSTRINGS, 0, 0, 0, 0, hwnd, idMenu(properties_list_id), self.instance, null);
		for ([_]win32.HWND{ self.categories_label, self.categories_list, self.properties_label, self.properties_list }) |control| _ = win32.SendMessageW(control, win32.WM_SETFONT, @intFromPtr(font), 1);
		const file_menu = win32.CreatePopupMenu();
		_ = win32.AppendMenuW(file_menu, win32.MF_STRING, menu_refresh_id, std.unicode.utf8ToUtf16LeStringLiteral("&Refresh\tF5"));
		_ = win32.AppendMenuW(file_menu, win32.MF_SEPARATOR, 0, null);
		_ = win32.AppendMenuW(file_menu, win32.MF_STRING, menu_exit_id, std.unicode.utf8ToUtf16LeStringLiteral("E&xit"));
		const edit_menu = win32.CreatePopupMenu();
		_ = win32.AppendMenuW(edit_menu, win32.MF_STRING, menu_copy_id, std.unicode.utf8ToUtf16LeStringLiteral("&Copy\tCtrl+C"));
		const menu_bar = win32.CreateMenu();
		_ = win32.AppendMenuW(menu_bar, win32.MF_POPUP, @intFromPtr(file_menu), std.unicode.utf8ToUtf16LeStringLiteral("&File"));
		_ = win32.AppendMenuW(menu_bar, win32.MF_POPUP, @intFromPtr(edit_menu), std.unicode.utf8ToUtf16LeStringLiteral("&Edit"));
		_ = win32.SetMenu(hwnd, menu_bar);
		const accels = [_]win32.ACCEL{ .{ .fVirt = win32.FVIRTKEY, .key = win32.VK_F5, .cmd = menu_refresh_id }, .{ .fVirt = win32.FVIRTKEY | win32.FCONTROL, .key = 'C', .cmd = menu_copy_id } };
		self.accelerators = win32.CreateAcceleratorTableW(&accels, accels.len);
		self.refreshCategories();
		_ = win32.SetFocus(self.categories_list);
	}

	fn onSize(self: *MainWindow, width: u16, height: u16) void {
		if (self.categories_list == null) return;
		const margin: i32 = 10;
		const label_height: i32 = 18;
		const half_width = @divTrunc(@as(i32, width) - margin * 3, 2);
		const list_top = margin + label_height;
		const list_height = @as(i32, height) - list_top - margin;
		const right_x = margin * 2 + half_width;
		_ = win32.MoveWindow(self.categories_label, margin, margin, half_width, label_height, 1);
		_ = win32.MoveWindow(self.categories_list, margin, list_top, half_width, list_height, 1);
		_ = win32.MoveWindow(self.properties_label, right_x, margin, half_width, label_height, 1);
		_ = win32.MoveWindow(self.properties_list, right_x, list_top, half_width, list_height, 1);
	}

	fn onCommand(self: *MainWindow, wparam: win32.WPARAM, lparam: win32.LPARAM) void {
		const id = wparam & 0xffff;
		if (lparam != 0) {
			if (id == categories_list_id and (wparam >> 16) == win32.LBN_SELCHANGE) {
				const index: i32 = @intCast(@as(isize, @bitCast(win32.SendMessageW(self.categories_list, win32.LB_GETCURSEL, 0, 0))));
				self.populateProperties(index);
			}
			return;
		}
		if (id == menu_refresh_id) {
			self.refreshCategories();
		} else if (id == menu_exit_id) {
			_ = win32.DestroyWindow(self.hwnd);
		} else if (id == menu_copy_id) {
			self.copySelectedProperty();
		} else if (id == @as(usize, @intCast(win32.IDCANCEL))) {
			_ = win32.DestroyWindow(self.hwnd);
		}
	}

	fn refreshCategories(self: *MainWindow) void {
		_ = win32.SendMessageW(self.categories_list, win32.LB_RESETCONTENT, 0, 0);
		_ = win32.SendMessageW(self.properties_list, win32.LB_RESETCONTENT, 0, 0);
		_ = self.arena.reset(.retain_capacity);
		const allocator = self.arena.allocator();
		var collected = std.ArrayList(CategoryItem).empty;
		for (provider_registry.providers) |get_items| {
			const provider_items = get_items(allocator) catch continue;
			collected.appendSlice(allocator, provider_items) catch continue;
		}
		std.mem.sort(CategoryItem, collected.items, {}, struct {
			fn lessThan(_: void, a: CategoryItem, b: CategoryItem) bool {
				return std.mem.lessThan(u8, a.label, b.label);
			}
		}.lessThan);
		self.items = collected.items;
		for (self.items) |item| {
			const wide = toWide(allocator, item.label) catch continue;
			_ = win32.SendMessageW(self.categories_list, win32.LB_ADDSTRING, 0, ptrToLparam(wide.ptr));
		}
		if (self.items.len > 0) {
			_ = win32.SendMessageW(self.categories_list, win32.LB_SETCURSEL, 0, 0);
			self.populateProperties(0);
		}
	}

	fn populateProperties(self: *MainWindow, category_index: i32) void {
		_ = win32.SendMessageW(self.properties_list, win32.LB_RESETCONTENT, 0, 0);
		if (category_index < 0 or category_index >= self.items.len) return;
		const allocator = self.arena.allocator();
		for (self.items[@intCast(category_index)].properties) |prop| {
			if (prop.value.len == 0) continue;
			const line = std.mem.concat(allocator, u8, &.{ prop.name, ": ", prop.value }) catch continue;
			const wide = toWide(allocator, line) catch continue;
			_ = win32.SendMessageW(self.properties_list, win32.LB_ADDSTRING, 0, ptrToLparam(wide.ptr));
		}
		if (win32.SendMessageW(self.properties_list, win32.LB_GETCOUNT, 0, 0) > 0) _ = win32.SendMessageW(self.properties_list, win32.LB_SETCURSEL, 0, 0);
	}

	fn copySelectedProperty(self: *MainWindow) void {
		if (win32.GetFocus() != self.properties_list) return;
		const index: i32 = @intCast(@as(isize, @bitCast(win32.SendMessageW(self.properties_list, win32.LB_GETCURSEL, 0, 0))));
		if (index == win32.LB_ERR) return;
		const len: isize = @bitCast(win32.SendMessageW(self.properties_list, win32.LB_GETTEXTLEN, @intCast(index), 0));
		if (len == win32.LB_ERR) return;
		const buffer = std.heap.page_allocator.alloc(u16, @as(usize, @intCast(len)) + 1) catch return;
		defer std.heap.page_allocator.free(buffer);
		_ = win32.SendMessageW(self.properties_list, win32.LB_GETTEXT, @intCast(index), ptrToLparam(buffer.ptr));
		if (win32.OpenClipboard(self.hwnd) == 0) return;
		defer _ = win32.CloseClipboard();
		_ = win32.EmptyClipboard();
		const byte_count = (@as(usize, @intCast(len)) + 1) * @sizeOf(u16);
		const mem = win32.GlobalAlloc(win32.GMEM_MOVEABLE, byte_count);
		if (mem == null) return;
		const dest = win32.GlobalLock(mem);
		if (dest == null) {
			_ = win32.GlobalFree(mem);
			return;
		}
		const dest_slice: [*]u16 = @ptrCast(@alignCast(dest));
		@memcpy(dest_slice[0 .. @as(usize, @intCast(len)) + 1], buffer);
		_ = win32.GlobalUnlock(mem);
		if (win32.SetClipboardData(win32.CF_UNICODETEXT, mem) == null) _ = win32.GlobalFree(mem);
	}
};
